src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs

```diff
@@ -459,26 +455,106 @@ static internal bool IsNonAttributedTypeValidForSerialization(Type type)
+        {
+            get
+            {
+                if (s_knownSerializableGenericTypeInfos == null)
+                {
+                    lock (s_knownSerializableGenericTypeInfosLock)
+                    {
+                        if (s_knownSerializableGenericTypeInfos == null)
+                        {
+                            s_knownSerializableGenericTypeInfos = new Dictionary<string, SerializableInfo>();
+                            InitializeKnownSerializableTypesFromStringList(s_knownSerializableGenericTypeNames, s_knownSerializableGenericTypeInfos);
+                        }
+                    }
+                }
+                return s_knownSerializableGenericTypeInfos;
+            }
+        }
+
+        private static object s_knownSerializableNonGenericTypeInfosLock = new object();
+        private static volatile IDictionary<string, SerializableInfo> s_knownSerializableNonGenericTypeInfos;
+        private static IDictionary<string, SerializableInfo> KnownSerializableNonGenericTypes
+        {
+            get
+            {
+                if (s_knownSerializableNonGenericTypeInfos == null)
+                {
+                    lock (s_knownSerializableNonGenericTypeInfosLock)
+                    {
+                        if (s_knownSerializableNonGenericTypeInfos == null)
+                        {
+                            s_knownSerializableNonGenericTypeInfos = new Dictionary<string, SerializableInfo>();
+                            InitializeKnownSerializableTypesFromStringList(s_knownSerializableNonGenericTypeNames, s_knownSerializableNonGenericTypeInfos);
+                        }
+                    }
+                }
+                return s_knownSerializableNonGenericTypeInfos;
+            }
+        }
+
+        private static void InitializeKnownSerializableTypesFromStringList(string[] listOfTypes, IDictionary<string, SerializableInfo> dict)
+        {
+            Debug.Assert(listOfTypes != null);
+            Debug.Assert(dict != null);
+            foreach (string str in listOfTypes)
+            {
+                SerializableInfo si = SerializableInfo.Parse(str);
+                dict.Add(si.TypeName, si);
+            }
+        }
+
         internal static bool IsKnownSerializableType(Type type)
         {
             // Applies to known types that DCS understands how to serialize/deserialize
             //
 
-            // Ajdust for generic type
-            if (type.GetTypeInfo().IsGenericType && !type.GetTypeInfo().IsGenericTypeDefinition)
+            if (type.GetTypeInfo().IsGenericType)
             {
-                type = type.GetGenericTypeDefinition();
+                if (!type.GetTypeInfo().IsGenericTypeDefinition)
+                {
+                    type = type.GetTypeInfo().GetGenericTypeDefinition();
+                }
+                return Enumerable.Contains(KnownSerializableGenericTypes.Keys, type.FullName);
```

Comment:
This could be replaced with `KnownSerializableGenericTypes.ContainsKey(type.FullName)`, which would avoid the enumerator allocation, and would be faster than looping through the `Keys` collection (which also avoids the one-time allocation of the Dictionary's `KeysCollection`).